lib/Component.spec.lua

```diff
@@ -207,6 +211,63 @@ return function()
 			end).to.throw()
 		end)
 
+		it("should throw when called in shouldUpdate", function()
+			local TestComponent = Component:extend("TestComponent")
+
+			local shouldUpdate
+
+			function TestComponent:init()
+				shouldUpdate = function()
+					self:shouldUpdate({}, {})
+				end
+			end
+
+			function TestComponent:render()
+				return nil
+			end
+
+			function TestComponent:shouldUpdate()
+				self:setState({
+					a = 1
+				})
+			end
+
+			local testElement = Core.createElement(TestComponent)
+
+			expect(function()
+				Reconciler.reify(testElement)
+				shouldUpdate()
+			end).to.throw()
```

Comment:
Since `shouldUpdate` is implemented entirely by the component, `_canSetState` won't be checked if we call `shouldUpdate` directly, right?

I wonder if we should call the component's `setState` instead of force updating, to hit paths like this one?

Reply:
This is true! It looks like this test case only passed to begin with when it re-entered `shouldUpdate`. I've fixed that now.